### PR TITLE
Create Character Count from training text

### PR DESCRIPTION
**Pull request by Shreeshrii**
Character frequency report 
Source: https://github.com/cmroughan/kraken_generated-data/blob/master/tools/count_chars.py

USAGE: count_chars.py <txt_file> | sort -n -r > <txt_file>.charcount

### CODE REVIEW

**Comment by bertsky** [comment]
Fixes #221 IINM

File: count_chars.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Source: https://github.com/cmroughan/kraken_generated-data/blob/master/tools/count_chars.py
```
Comment (kba) [inline count_chars.py:4]: Note that this is GPL code. Not sure about the consequences.

**Review by kba — APPROVED** [review]
A useful script. Maybe add a sentence about it to the README so people know its there and/or a Makefile target for this?

**Review by bertsky — CHANGES_REQUESTED** [review]
Useful indeed, but I wonder why they sort outside on the shell if we already run Python. 

Also, it would help to offer a rule for the makefile already.

Besides, in https://github.com/tesseract-ocr/tesstrain/pull/260/commits/314e799e8d5457d9792ee7e6b62b60b93036e64f I proposed a similar functionality (only using shell means, i.e. `grep -o . | sort | uniq -c | sort -rn`) – it does not show codepoint names via `unicodedata.name`, but apart from that should be the same).

**Comment by Shreeshrii** [comment]
@bertsky I copied the script from another repo and added an external sort as a quick hack to get the char count. Please feel free to modify as needed.

File: count_chars.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import unicodedata
+import sys, getopt
+
+def main(argv):
+    txt_file = ''
+    try:
+        opts, args = getopt.getopt(argv,"h")
+    except getopt.GetoptError:
+        print('USAGE: count_chars.py <txt_file>')
```
Comment (bertsky) [inline count_chars.py:14]:
```suggestion
        print(usage)
```

File: count_chars.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        sys.exit(2)
+    for opt, arg in opts:
+        if opt in ('-h','--help'):
+            print('USAGE: count_chars.py <txt_file> | sort -n -r > <txt_file>.charcount')
```
Comment (bertsky) [inline count_chars.py:18]:
```suggestion
            print(usage)
```

File: count_chars.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+            sys.exit()
+    for arg in args:
+        txt_file = arg
+    
+    inFile = open(txt_file)
+    rawText = inFile.read()
+    inFile.close()
```
Comment (bertsky) [inline count_chars.py:25]:
```suggestion
    rawText = ''
    for arg in args:
        with open(arg) as fd:
            rawText.append(fd.read())
    if not args:
        rawText.append(sys.stdin.read())
```

File: count_chars.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    chars = {}
+    for char in rawText:
+        if char not in chars:
+            chars[char] = 1
+        else:
+            chars[char] +=1
+
+    keys = list(chars.keys())
+    keys.sort()
```
Comment (bertsky) [inline count_chars.py:35]:
```suggestion
    if '-n' in opts or '--numeric' in opts:
        keys = sorted(chars.keys(), key=chars.get)
    else:
        # apply system locale
        locale.setlocale(locale.LC_ALL, '')
        keys = sorted(chars.keys(), key=locale.strxfrm)
```

File: count_chars.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+# Source: https://github.com/cmroughan/kraken_generated-data/blob/master/tools/count_chars.py
+
+import unicodedata
+import sys, getopt
+
+def main(argv):
+    txt_file = ''
+    try:
+        opts, args = getopt.getopt(argv,"h")
```
Comment (bertsky) [inline count_chars.py:12]:
```suggestion
        opts, args = getopt.getopt(argv, "hn", ["help", "numeric"])
```

**Review by bertsky — COMMENTED** [review]
This adds sorting (by current locale, or optionally by frequency), as well as multi-file input.

Still missing: makefile rule, using `$(ALL_GT)`. (I cannot use Github suggestions for this.)

File: count_chars.py (hunk 1 above)
Comment (stweil) [reply count_chars.py:4]: Using GPL code requires changing the license for tesstrain to GPL. I don't think that we should do this, so GPL code cannot be used here.

**Review by stweil — CHANGES_REQUESTED** [review]
We cannot use GPL code in tesstrain with Apache license.

**Comment by zdenop** [comment]
Please make a description of input data and needed output, so those of us how did not have a look at the file can create a new script under Apache license.

**Comment by bertsky** [comment]
> We cannot use GPL code in tesstrain with Apache license.

I don't think this short script meets the threshold of originality, though. After all, it just counts characters of files in Python. My above suggestions already changed most of the script's lines to make it more useful and makefile-workable. Just apply them and strip the Kraken reference. (I am not contesting Kraken's originality, only that single file's.)

**Comment by bertsky** [comment]
> Also, it would help to offer a rule for the makefile already.
> 
> Besides, in [314e799](https://github.com/tesseract-ocr/tesstrain/commit/314e799e8d5457d9792ee7e6b62b60b93036e64f) I proposed a similar functionality (only using shell means, i.e. `grep -o . | sort | uniq -c | sort -rn`) – it does not show codepoint names via `unicodedata.name`, but apart from that should be the same).

Or abandon this PR and just merge #260.

**Comment by zdenop** [comment]
Here is my python solution without need of extra tools. I did not implement reading form `stdin` as I do not see it usage in "make training"....

Usage: `python3 count_chars.py data/foo/all-gt`

```py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


import sys
import unicodedata
from collections import Counter, OrderedDict


def show_char_frequency(string):
    string = string.replace(" ", "")
    dic = OrderedDict(Counter(string).most_common())
    for char in dic:
        print(f"{char}\t{dic[char]}\t{unicodedata.name(char)}")


def read_file(filename):
    with open(filename, encoding="utf-8", mode="rt") as fd:
        text_lines = fd.read().strip().split("\n")
    return " ".join(text_lines)


def main():
    if len(sys.argv) < 2:
        print(f"USAGE: {sys.argv[0]} <txt_file>")
        return 1

    filename = sys.argv[1]
    string = read_file(filename)
    show_char_frequency(string)
    return 0


if __name__ == "__main__":
    main()
```

File: count_chars.py (hunk 1 above)
Comment (stweil) [reply count_chars.py:4]: @cmroughan, are you the author of that code, and would you agree to use it here under the Apache 2.0 License?

Comment (cmroughan) [reply count_chars.py:4]: I am, and I wouldn't mind its use here under Apache 2.0. I will follow up with a pull request.

**Comment by stweil** [comment]
Commit 29d394b7253c9f933a7fdf57f553d305576f9a5d merged the modifications (based on the original code) which were made in this pull request.

**Pull request closed without merging**